sql/join.go

```diff
@@ -0,0 +1,172 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const joinBatchSize = 100
+
+// An indexJoinNode implements joining of results from an index with the rows
+// of a table. The index side of the join is pulled first and the resulting
+// rows are used to lookup rows in the table. The work is batched: we pull
+// joinBatchSize rows from the index and use the primary key to construct spans
+// that are looked up in the table.
+type indexJoinNode struct {
+	index            *scanNode
+	table            *scanNode
+	primaryKeyPrefix proto.Key
+	colIDtoRowIndex  map[ColumnID]int
+	err              error
+}
+
+func makeIndexJoin(indexScan *scanNode) (*indexJoinNode, error) {
+	// Copy the index scan node into a new table scan node and reset the fields
+	// that were set up for the index scan.
+	table := &scanNode{}
+	*table = *indexScan
+	table.index = &table.desc.PrimaryIndex
+	table.spans = nil
+	table.reverse = false
+	table.isSecondaryIndex = false
+	table.initOrdering()
+
+	// We want to the index scan to keep the same render target indexes for
+	// columns which are part of the primary key or part of the index. This
+	// allows sorting (which was calculated based on the output columns) to be
+	// avoided if possible.
+	colIDtoRowIndex := map[ColumnID]int{}
+	for _, colID := range table.desc.PrimaryIndex.ColumnIDs {
+		colIDtoRowIndex[colID] = -1
+	}
+	for _, colID := range indexScan.index.ColumnIDs {
+		colIDtoRowIndex[colID] = -1
+	}
+
+	// Rebuild the render targets for the index scan by looping over the render
+	// targets for the table. Any referenced column that is part of the index or
+	// the primary key is kept in its current location. Any other render target
+	// is replaced with "1".
+	indexScan.render = nil
+	for _, render := range table.render {
+		switch t := render.(type) {
+		case *qvalue:
+			if _, ok := colIDtoRowIndex[t.col.ID]; ok {
+				colIDtoRowIndex[t.col.ID] = len(indexScan.render)
+				indexScan.render = append(indexScan.render, t)
+				continue
+			}
+		}
+		indexScan.render = append(indexScan.render, parser.DInt(1))
+	}
+	for colID, index := range colIDtoRowIndex {
+		if index == -1 {
+			col, err := table.desc.FindColumnByID(colID)
+			if err != nil {
+				return nil, err
+			}
+			colIDtoRowIndex[col.ID] = len(indexScan.render)
+			indexScan.render = append(indexScan.render, indexScan.getQVal(*col))
+		}
+	}
```

Comment:
FYI, I added another commit which made the above a bit more complicated. The upshot is that we can avoid sorting when the index scan outputs rows in the desired order.